docs/proposals/20210222-kubelet-authentication.md

```diff
@@ -0,0 +1,331 @@
+
+The kubelet serving certificate rotation daemon runs as a systemd or windows service manager service on the node. It checks for the existence of the serving certificate files, their validity (e.g. not self signed), or request a new serving certificate. Once the new serving certificate is written, kubelet can be restarted.
+
+#### Core Specification
+- Core Cluster API MUST provide the following implementations of CSRs and signers:
+    - `cluster.x-k8s.io/kube-apiserver-client-kubelet-insecure` which implement an “Always Allow” type signer that provides equivalent security to Cluster API v1alpha3. This is only to be used for providers where no secure mechanism exists.
+    - `cluster.x-k8s.io/kube-apiserver-client-kubelet-tpm` and `cluster-x-k8s-io/kubelet-serving-tpm`
+        - Will implement TPM-based certificate signers and requesters based on the cloud-provider-gcp implementation.
+        - However, since the mechanism for retrieving endorsement keys varies across platforms, the TPM signer will additionally require a provider specific mechanism to provide the TPM Endorsement Key.
+
+#### Provider Specification
+
+#### All providers
+- All providers MUST insert a ProviderID within the KubernetesNodeProviderIdentifierOID extension attribute of the CSR
+- All signer names MUST be filled in by the provider’s controller in InfraCluster.Status.KubeletClientCertificateSigner and InfraCluster.Status.KubeletServingCertificateSigner
+
+#### Insecure providers
+- An insecure provider CANNOT implement certificate rotation or kubelet serving certificate signing
+- InfraCluster.Status.KubeletClientCertificateSigner MUST be set to cluster.x-k8s.io/kube-apiserver-client-kubelet-insecure
+- An insecure provider MUST use the cluster.x-k8s.io/kube-apiserver-client-kubelet-insecure signer
+
+#### Secure providers
+- A secure provider MUST implement certificate rotation and kubelet server certificate signing
+- A provider must register signers of:
+    - `cluster-x-k8s-io/kube-apiserver-client-kubelet-<provider>`
+    - `cluster-x-k8s-io/kubelet-serving-<provider>`
+- A secure provider MUST implement a secure attestation mechanism, based upon PEM-encoded blocks within the Certificate Signing Request.
+- Where a secure provider’s attestation mechanism does not include a nonce or timestamp to protect against replay attacks, the mechanism MUST implement a secondary time-limited attestation (e.g. AWS Instance Identity document + AWS HMACv4 signature).
+- A provider’s signer MUST run on the management cluster
+
+#### TPM based providers
+- A TPM provider MUST use the following certificate signers
+    - `cluster-x-k8s-io/kube-apiserver-client-kubelet-tpm`
+    - `cluster-x-k8s-io/kubelet-serving-tpm`
+- A TPM provider MUST annotate new CSRs as follows:
+    - Key: cluster-x-k8s-io/tpm-endorsement-key
+    - Value: Platform-specific endorsement key (e.g., retrieved from GCP Shielded VM API or VMware vCenter)
+
+#### Kubeadm
```

Comment:
i think, we actually might have another problem here because kubeadm hardcodes the kubelet cert rotation to "true". we don't want to, but the problem is that there is no way to tell if the users specified it explicitly as "false".
(the field is a `bool` and `*bool` and the default of any empty go struct is `false`).

if kubeadm continues to manage the kubeletconfiguration in this proposal, i'd assume cert rotation must be set to `false`?
adding support for kubeletconfiguration patches per node is a workaround, but CAPI cannot use them declaratively yet because they are not in the kubeadm config.

Reply:
yeah we will need to disable client cert rotation. does kubeadm provide a possibility to apply patches. If not, we'll workaround it in the bootstrap script

Reply:
patches for the KubeletConfiguration are not supported.
currently joining nodes inherit a global KubeletConfiguration passed to kubeadm init.

the whole `*bool` problem makes it difficult to solve.

this can be done in the following way:
- kubeadm init is called
- the /var/lib/kubelet/config.yaml on the init node is modified to disable rotate
- the kubelet there is restarted
- the kube-system/kubelet-config-x.yy is patched to disable rotation too so that joining nodes don't have it.

Reply:
What are the side effects of not having kubelet certificate rotation in place? The point @neolit123 raised here seems it warrants some details in the proposal